<a href="https://colab.research.google.com/github/MiguelAngelRC1115/Analitica-de-Negocios/blob/main/HeartDiseaseMiguelAngelRoa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Caso de Estudio**

Una entidad de salud desea implementar un modelo de árbol de decisión para pronosticar si un paciente desarrollará una enfermedad cardíaca con base en sus características clínicas. Para este proceso se utilizan las variables:

**Disease (Enfermedad):** indica si el paciente desarrolla (1) o no (0) una enfermedad cardíaca.

**Age (Edad):** edad del paciente en años.

**Systolic (Presión sistólica):** Presión arterial sistólica (mmHg). Niveles elevados están asociados con mayor riesgo cardiovascular.

**Diastolic (Presión diastólica):** Presión arterial diastólica (mmHg). Refleja la presión cuando el corazón está en reposo entre latidos.

**BMI (IMC):** Índice de Masa Corporal (kg/m²). Indicador de composición corporal.

0. Se procede con la carga de las librerías de trabajo

In [ ]:
import numpy as np                              # Librería para operaciones
import pandas as pd                             # Para el análisis de datos
import matplotlib.pyplot as plt                 # Para generación de gráficos
from sklearn.tree import DecisionTreeClassifier # Para construir el modelo de árbol de decisión
from sklearn.metrics import confusion_matrix    # Para la matriz de confusión

1. Se procede con la carga de los datos de trabajo

In [ ]:
nxl = ('/content/2. HeartDisease.xlsx') #Ruta del excel
XDB = pd.read_excel(nxl, sheet_name=0) #Se lee el excel
XD = XDB[['Age', 'Systolic', 'Diastolic', 'BMI']] #Se seleccionan las columnas de las variables que vamos a usar
yd = XDB['Disease'] #Se selecciona la columna de la variable a predecir
display(XD) #Se visualiza la base de datos


,Age,Systolic,Diastolic,BMI
0,44,112,111,17
1,55,128,90,27
2,47,131,94,26
3,31,151,104,17
4,65,148,117,17
...,...,...,...,...
175,43,119,76,25
176,63,107,113,40
177,47,116,61,27
178,70,110,88,24


2. Se procede con la implementación del modelo del árbol

In [ ]:
mar = DecisionTreeClassifier(criterion='gini', max_depth=4)
mar.fit(XD, yd) #El modelo busca relacion entrada - salida

ydp = mar.predict(XD) #Lo que pronostica el modelo

# Se construye la matriz de confusión
cm = confusion_matrix(yd, ydp)
display(cm)
VN = cm[0, 0]; FP = cm[0, 1]; FN = cm[1, 0]; VP = cm[1, 1]

# Métricas de desempeño
EX = (VP + VN) / len(XD)     # 1. Exactitud: comportamiento general del modelo
print('La exactitud es: ', EX)
sen = VP / (VP + FN)           # 2. Sensibilidad: como se comporta pronosticando enfermos
print('La sensibilidad es: ', sen)
spe = VN / (VN + FP)           # 3. Especificidad: como se comporta pronosticando sanos
print('La especificidad es: ', spe)
pre = VP / (VP + FP)           # 4. Precisión positiva: de los pronosticados como enfermos, cuántos si lo son
print('La precisión es: ', pre)
preneg = VN / (VN + FN)        # 5. Precisión negativa: de los pronosticados como sanos, cuántos si lo son
print('La precisión negativa es: ', preneg)

array([[63, 10],
       [20, 87]])

La exactitud es:  0.8333333333333334
La sensibilidad es:  0.8130841121495327
La especificidad es:  0.863013698630137
La precisión es:  0.8969072164948454
La precisión negativa es:  0.7590361445783133


3. Despliegue del arbol de decision

In [ ]:
from sklearn.tree import export_graphviz #exporta datos a un grafico
from pydotplus import graph_from_dot_data #graficador

vs=["Age","Systolic","Diastolic","BMI"] #titulos del arbol
dot_data=export_graphviz(mar,feature_names=vs) #exportar de numeros a grafico
graph=graph_from_dot_data(dot_data) #hacemos el grafico
graph.write_png('arbol.png')

True

4. Pronostico nuevos pacientes - hoja 2

In [ ]:
XDB2 = pd.read_excel(nxl, sheet_name=1) #Se cargan los datos de los nuevos pacientes
XD2 = XDB2[['Age', 'Systolic', 'Diastolic', 'BMI']] #Se seleccionan las variables otra vez
ydp2 = mar.predict(XD2) #Se generan los pronosticos para los nuevos pacientes
prob2 = mar.predict_proba(XD2) #Se calculan las probabilidades
print("Pronóstico para nuevos pacientes:") #Se muestra el diagnostico
for i in range(len(ydp2)):
    estado = "Con Enfermedad Cardíaca" if ydp2[i] == 1 else "Sin Enfermedad Cardíaca"
    print(f"  Paciente {i+1}: {estado} | Prob. Sano: {prob2[i][0]:.1%} | Prob. Enfermo: {prob2[i][1]:.1%}")


Pronóstico para nuevos pacientes:
  Paciente 1: Sin Enfermedad Cardíaca | Prob. Sano: 75.8% | Prob. Enfermo: 24.2%
  Paciente 2: Sin Enfermedad Cardíaca | Prob. Sano: 60.0% | Prob. Enfermo: 40.0%
  Paciente 3: Sin Enfermedad Cardíaca | Prob. Sano: 75.8% | Prob. Enfermo: 24.2%
  Paciente 4: Con Enfermedad Cardíaca | Prob. Sano: 6.2% | Prob. Enfermo: 93.8%
  Paciente 5: Con Enfermedad Cardíaca | Prob. Sano: 0.0% | Prob. Enfermo: 100.0%
  Paciente 6: Con Enfermedad Cardíaca | Prob. Sano: 6.2% | Prob. Enfermo: 93.8%
  Paciente 7: Con Enfermedad Cardíaca | Prob. Sano: 0.0% | Prob. Enfermo: 100.0%
  Paciente 8: Con Enfermedad Cardíaca | Prob. Sano: 0.0% | Prob. Enfermo: 100.0%


**Análisis de resultados**

De la base de datos se pueden observar un total de 73 pacientes clasificados como sanos, mientras que 107 pacientes presentan enfermedad cardíaca, para un total de 180 registros analizados.

Con respecto a las métricas de desempeño, el modelo obtuvo una exactitud del 83.33%, lo que refleja un comportamiento general sólido al clasificar correctamente tanto pacientes enfermos como sanos. La sensibilidad del 81.31% indica que el modelo identifica correctamente a más del 81% de los pacientes que realmente padecen la enfermedad cardíaca, mientras que la especificidad del 86.30% muestra que es aún más preciso al reconocer a los pacientes sanos. Ambas métricas superan el umbral mínimo del 75% recomendado para modelos de clasificación clínica, lo que valida el desempeño general del árbol de decisión.
La precisión positiva del 89.69% es la métrica más destacada del modelo, significando que de cada 10 pacientes pronosticados como enfermos, casi 9 efectivamente lo son, lo cual reduce considerablemente los falsos positivos y otorga alta confiabilidad al diagnóstico positivo. La precisión negativa del 75.90%, aunque es la métrica más baja del conjunto, se mantiene por encima del umbral mínimo, indicando que el modelo también tiene un desempeño aceptable al confirmar los diagnósticos negativos. En conjunto, el modelo demuestra un balance adecuado entre todas las métricas, siendo especialmente robusto en la identificación de casos positivos, lo que lo hace pertinente para su aplicación en un contexto de clasificación clínica de riesgo cardiovascular.

Con respecto al árbol de decisión, este usa el índice Gini. La variable más importante es la edad, lo que confirma que es el principal factor para predecir la enfermedad cardíaca en esta población. La primera división separa a los pacientes de 54.5 años o menos (122 casos) de los mayores (58 casos); este último grupo ya muestra un riesgo alto, lo que sugiere que superar esa edad es en sí una señal importante.

Explorando el lado derecho del árbol (Age > 54.50), se identifica un nodo puro destacable: pacientes con Age > 54.50 y Age > 69.50 presentan una probabilidad de enfermedad cardíaca del 0% (2 casos, Gini=0.00), lo que constituye una regla de negación. Por otro lado, el nodo con Age > 54.50, Age <= 69.50 y Diastolic > 85.50 representa un nodo puro de aprobación con el 100% de probabilidad de desarrollar enfermedad cardíaca (35 casos, Gini=0.00). Esto significa que pacientes mayores de 54 años con presión diastólica superior a 85.5 mmHg tienen certeza absoluta de diagnóstico positivo según el modelo.

Respecto al pronóstico de nuevos pacientes, 6 de los 8 nuevos casos son clasificados como pacientes con enfermedad cardíaca. Los pacientes 5, 7 y 8 presentan una probabilidad del 100% de desarrollar la enfermedad. Los pacientes 1 y 3 son los únicos pronosticados como sanos, con una probabilidad del 73.5% de no padecer la enfermedad